geoutils/georaster.py

```diff
@@ -357,23 +366,54 @@ def reproject(self, dst_crs, dst_size=None, dst_bounds=None, dst_res=None,
         :dtype nodata: int, float, None
         :param resampling: A rasterio Resampling method
         :dtype resample: rio.warp.Resampling object
-        :param **kwargs: additional keywords are passed to rasterio.warp.reproject. Use with caution.         
+        :param **kwargs: additional keywords are passed to rasterio.warp.reproject. Use with caution.
 
         :returns: Raster
         :rtype: Raster
-
         """
 
-        # Check input arguments
+        # Check that either dst_ref or dst_crs is provided
+        if dst_ref is not None:
+            if dst_crs is not None:
+                raise ValueError("Either of `dst_ref` or `dst_crs` must be set. Not both.")
+        else:
+            if dst_crs is None:
+                raise ValueError("One of `dst_ref` or `dst_crs` must be set.")
+            
+        # Case a raster is provided as reference
+        if dst_ref is not None:
+
+            # Check that dst_ref type is either str, Raster or rasterio data set
+            if isinstance(dst_ref, Raster):
+                ds_ref = dst_ref.ds
```

Comment:
What is the rationale behind getting the crs,width, height, bounds from the underlying rasterio dataset? All these attributes are copied into the Raster object.

I've found an edge case where this is "important". When doing multiprocessing such as with dask, Rasters with opened rasterio.DatasetReaders cannot be pickled for transfer to a compute node. I found this when trying to pass a Raster containing georeferencing information as an argument for clipping/reprojecting several years of data, where each year is a multiprocessing task. 

If this line were changed to `ds_ref = dst_ref`, then all extra that would be needed would be for the user to run `myim.ds.close()` before dispatching a Raster to a compute node.

Reply:
True, I didn't realize that ds_ref = dst_ref would work too with the correct attributes transferred to the Raster object (although the user can choose not to load them with the attr argument in __init__...
I'm not sure to understand the end message in the end. Should we change this line or it's better not to? If not changing, should I add a comment to the code to explain it?

Reply:
Sorry, I should have been clearer! :)
1) Actually, all the attrs listed in `default_attrs` always get loaded. See the `_read_attrs` method: if the `attrs` keyword is provided by the user, any items in `default_attrs`which are missing from the user-provided list get added automatically.
2) Because of (1), I think that it's safe and useful to change this line to `ds_ref = dst_ref`.

Reply:
Regarding my comment above,

> Rasters with opened rasterio.DatasetReaders cannot be pickled for transfer to a compute node.

I've opened an associated issue regarding closing these readers (#50 ).

Reply:
I would like to add a comment in the code to remember this. Is that good so: "# Preferably use Raster instance to avoid rasterio data set to remain open. See PR #45"?

Reply:
Just pushed the changes. I also modified the case where dst_ref is str.

Reply:
> Rasters with opened rasterio.DatasetReaders cannot be pickled for transfer to a compute node.

Yes, this is a pain indeed. I kind of manually circumvent this issue generally. We'll see how to best implement through the class...